src/Utilities/CleverDicts.jl

```diff
@@ -0,0 +1,198 @@
+    # Perform this `haskey` check up front to detect setting with keys that are
+    # invalid (i.e., have already been deleted). You can only call setindex!
+    # with a key obtained from `new_key` that hasn't been deleted.
+    if !haskey(c, key)
+        throw(KeyError(key))
+    elseif c.dict === nothing
+        @assert c.vector !== nothing
+        c.vector[key_to_index(key)] = val
+    else
+        c.dict[key] = val
+    end
+    return val
+end
+
+struct LinearIndex
+    i::Int
+end
+
+function Base.getindex(c::CleverDict{K, V}, index::LinearIndex)::V where {K, V}
+    if !(1 <= index.i <= length(c))
+        throw(KeyError(index))
+    end
+    # Get the `index` linear element. If `c.vector` is currently `nothing`
+    # (i.e., there has been a deletion), rebuild `c.vector`. This is a
+    # trade-off: We could ensure `c.vector` is always updated, but this requires
+    # a `splice!` in `delete!`, making deletions costly. However, it makes this
+    # `getindex` operation trival because we would never have to rebuild the
+    # vector.
+    # The current implemented approach offers quick deletions, but an expensive
+    # rebuild the first time you query a `LinearIndex` after a deletion or a new
+    # key is added. Once the rebuild is done, there are quick queries until the
+    # next deletion or addition. Thus, the worst-case is a user repeatedly
+    # deleting a key and then querying a LinearIndex (e.g., getting the MOI
+    # objective function).
+    if c.vector === nothing
+        c.vector = Vector{V}(undef, length(c))
+        for (i, val) in enumerate(values(c.dict))
+            c.vector[i] = val
+        end
+    end
+    return c.vector[index.i]::V
+end
+
+function Base.delete!(c::CleverDict{K, V}, key::K)::Nothing where {K, V}
+    if c.dict === nothing
+        c.dict = OrderedCollections.OrderedDict{K, Union{Nothing, V}}()
+        for (i, info) in enumerate(c.vector)
+            c.dict[index_to_key(K, i)] = info
+        end
+    end
+    delete!(c.dict, key)
+    c.vector = nothing
+    return
+end
+
+function Base.length(c::CleverDict)::Int
+    return c.dict == nothing ? length(c.vector) : length(c.dict)
+end
+
+Base.haskey(::CleverDict, key) = false
+function Base.haskey(c::CleverDict{K, V}, key::K)::Bool where {K, V}
+    if c.dict === nothing
+        return 1 <= key_to_index(key) <= length(c.vector)
+    else
+        return haskey(c.dict, key)
+    end
+end
+
+function Base.iterate(
+    c::CleverDict{K, V}
+)::Union{Nothing, Tuple{Pair{K, V}, Int}} where {K, V}
+    if length(c) == 0
+        return nothing
+    elseif c.dict !== nothing
+        return iterate(c.dict)
+    else
+        @assert c.vector !== nothing
+        key = index_to_key(K, 1)
+        return key => c.vector[1], 2
+    end
+end
+
+function Base.iterate(
+    c::CleverDict{K, V}, s::Int
+)::Union{Nothing, Tuple{Pair{K, V}, Int}} where {K, V}
+    if s > length(c)
```

Comment:
This could be used when the dict is not nothing which would not make sense

Reply:
Or you rely on the fact that the state of the dict is always smaller than its length